Bug 1563169 - migrate PRs logic to use RelEng taskgraph instead

MihaiTabara opened the pull request:
This is an attempt to split the awesome work that @mitchhentges did in https://github.com/mozilla/application-services/pull/1945/ and land it in gradual phases:
* first landing PRs and iterating
* then landing the master pushes and iterating
* take care of github releases
* cleanup of old existing logic of `decisionlib`

This PRs tracks the first phase, taking care of PRs.

MihaiTabara commented:
I cherry-picked the PR related work from @mitchhentges' https://github.com/mozilla/application-services/pull/1945/. I kept some of the bits common with `master` pushes since `[ci full]` is what a master push will do.

My proposition is to let this bake for a few days on PRs to spot any potential issues, and only add follow-up with `master` pushes. Github releases after that. 

Pinging @JohanLorenzo @eoger for r? Also CC @badboy 

* normal PR CI - https://firefox-ci-tc.services.mozilla.com/tasks/groups/aELvqAoXRD60v1SJSfYuVw
* [ci skip] CI - https://firefox-ci-tc.services.mozilla.com/tasks/groups/X_xsoIPBRCmf3H_3a9cb8g
* [ci full] CI - https://firefox-ci-tc.services.mozilla.com/tasks/groups/MXIq9ZLNREyYqJlHOgQWZw

I've compared each of the individual tasks and found differences but functionality-wise, they are the same. 

Current issues that I have:
* normal PR still schedules the toolchain tasks and docker image, for each commit I do. I was expecting them to be re-used, unless changed
* `ci skip` still schedules `build-docker-image` for some reason, even though it shouldn't
* `ci full` fails for some of the modules. @eoger - do the errors from https://firefox-ci-tc.services.mozilla.com/tasks/LDYmuv7_Q7ee5ghs_6VIEg/runs/0/logs/https%3A%2F%2Ffirefox-ci-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FLDYmuv7_Q7ee5ghs_6VIEg%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log#L959 sound familiar?

JohanLorenzo commented on taskcluster/app_services_taskgraph/transforms/script_to_run_task.py:
```diff
@@ -0,0 +1,30 @@
+        "/bin/bash",
+        "--login",
+        "-c",
+        "cat <<'SCRIPT' > ../script.sh && bash -e ../script.sh\n{}\nSCRIPT".format(script)
```
Blocker: I remember Mitch talked to me about this hack. We've made good progress since this code was first written. We can re-use the `run-commands` logic that was implemented in mobile projects.

Here's the logic: https://github.com/mozilla-mobile/fenix/blob/21606bf94f2cd2876841f904acf69ee450422c98/taskcluster/fenix_taskgraph/job.py#L51-L67
Here's how to use it: https://github.com/mozilla-mobile/fenix/blob/d719823b2f0a091224b9066006091469bd531754/taskcluster/ci/ui-test/kind.yml#L28-L31

I don't feel super confident in using this here-document. I foresee some shell quoting problems.

MihaiTabara replied:
Noted, will do.

JohanLorenzo replied:
Chatted over Slack. Okay to tackle this in a follow up, due to the complexity of the task.

MihaiTabara replied:
Addressed in https://github.com/mozilla/application-services/pull/3113

JohanLorenzo commented on taskcluster/ci/lint/kind.yml:
```diff
@@ -0,0 +1,29 @@
+    description: 'detekt'
+    run:
+      using: run-task
+      cwd: '{checkout}'
```
Nit: this can be bubble up in `job-defaults`

MihaiTabara replied:
Good call, I moved both `using` and `cwd` to `job-defaults`.

JohanLorenzo commented on taskcluster/docker/linux/Dockerfile:
```diff
@@ -0,0 +1,168 @@
+
+# We use this specific version because our decision task also runs on this one.
+# We also use that same version in decisionlib.py
+FROM ubuntu:bionic-20180821
```
Followup in another PR: Bump this outdated image.

MihaiTabara replied:
Noted in follow-up list.

JohanLorenzo commented on taskcluster/docker/linux/Dockerfile:
```diff
@@ -0,0 +1,168 @@
+
+RUN pip3 install --upgrade pip
+RUN pip3 install \
+    'taskcluster>=4,<5' \
```
Follow up in another PR: Use the latest taskcluster client.

MihaiTabara replied:
Noted, added to my todo-list for follow-ups.

JohanLorenzo commented on taskcluster/docker/linux/Dockerfile:
```diff
@@ -0,0 +1,168 @@
+ENV ANDROID_SDK_HOME /builds/worker/android-sdk
+ENV PATH ${PATH}:${ANDROID_SDK_HOME}/tools:${ANDROID_SDK_HOME}/tools/bin:${ANDROID_SDK_HOME}/platform-tools:/opt/tools:${ANDROID_SDK_HOME}/build-tools/${ANDROID_BUILD_TOOLS}
+
+RUN curl -sfSL --retry 5 --retry-delay 10 https://dl.google.com/android/repository/sdk-tools-linux-${ANDROID_SDK_VERSION}.zip > sdk.zip \
```
Follow up: https://bugzilla.mozilla.org/show_bug.cgi?id=1622339

MihaiTabara replied:
Noted, added to my list.

JohanLorenzo commented on taskcluster/scripts/toolchain/cross-compile-setup.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
I don't feel qualified to review this setup. I guess it's the right one.

MihaiTabara replied:
Me neither. As far as I can tell, copy-paste from decisionlib so it should be fine O_O

JohanLorenzo commented on taskcluster/scripts/toolchain/rustup-setup.sh:
```diff
@@ -0,0 +1,28 @@
+    RUSTUP_VERSION='1.18.3'; \
+    RUSTUP_SHA256='a46fe67199b7bcbbde2dcbc23ae08db6f29883e260e23899a88b9073effc9076'; \
+    curl -sfSL --retry 5 --retry-delay 10 -O "https://static.rust-lang.org/rustup/archive/${RUSTUP_VERSION}/${RUSTUP_PLATFORM}/rustup-init"; \
+    echo "${RUSTUP_SHA256} *rustup-init" | sha256sum -c -; \
```
👍 for checking checksums

JohanLorenzo commented on taskcluster/app_services_taskgraph/target_tasks.py:
```diff
@@ -0,0 +1,34 @@
+    via the `run_on_projects` attributes."""
+    def filter(task):
+        return filter_for_tasks_for(task, parameters) \
+            and task.attributes.get("run-on-pr-type", "all") in ("full-ci", "all")
```
Question: should the `normal-ci` jobs be part of the full run too? If they already are, how does it work? 🤔

JohanLorenzo commented on taskcluster/app_services_taskgraph/transforms/module_build.py:
```diff
@@ -0,0 +1,44 @@
+@transforms.add
+def release_upload_symbols(config, tasks):
+    for task in tasks:
+        if config.params["tasks_for"] == u"github-release":
```
Nit: This PR doesn't take `github-release`s into account. Let's remove this logic from the PR. I don't think this is the right way to go. In my opinion, we should have 2 `release-type`s: a regular one and a release one. The modifications done below can be bubbled up to `kind.yml`. See android-components, for example: https://github.com/mozilla-mobile/android-components/blob/7e38ed297b76e9cdc42b88e2af5a2cdad20af146/taskcluster/ci/build/kind.yml#L21-26

JohanLorenzo commented on taskcluster/app_services_taskgraph/transforms/module_build.py:
```diff
@@ -0,0 +1,44 @@
+        task["worker"]["artifacts"] = artifacts = []
+
+        for publication in module_info["publications"]:
+            primary_extensions = (".pom", ".aar", "-sources.jar") if publication["type"] == "aar" else (".pom", ".jar")
```
Followup: we can reuse the logic that a-c implemented https://github.com/mozilla-mobile/android-components/blob/8da8237bf9ccf1cf43088b995db5d4ca02c28a69/taskcluster/ac_taskgraph/build_config.py#L37

MihaiTabara replied:
Done in https://github.com/mozilla/application-services/pull/3113/files

MihaiTabara commented on .taskcluster.yml:
```diff
@@ -1,23 +1,23 @@
   $let:
+    taskgraph:
+      # TODO: to update this to production taskgraph once Tom's toolchain-script
+      # support gets added in https://phabricator.services.mozilla.com/D66709
```
I've got a new version working. Will push to taskgraph before landing the PR and adjusting here too.

MihaiTabara commented on taskcluster/app_services_taskgraph/transforms/out_of_docker_setup.py:
```diff
@@ -0,0 +1,82 @@
+def _cross_compile(config, tasks):
+    for task in tasks:
+        script = task["worker"]["script"]
+        task["worker"]["script"] = CROSS_COMPILE_SETUP + "\n" + script
```
Okay, will do.

JohanLorenzo replied:
Chatted over Slack. Okay to tackle this in a follow up, due to the complexity of the task.

MihaiTabara replied:
Addressed in https://github.com/mozilla/application-services/pull/3113

MihaiTabara commented on taskcluster/app_services_taskgraph/loader/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    for _, tasks in groups.iteritems():
+        tasks.extend([
+            task for task in tasks_for_all_components
+        ])
```
Done

MihaiTabara commented on taskcluster/app_services_taskgraph/loader/single_dep.py:
```diff
@@ -0,0 +1,32 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
I went again over the codebase and turns out we're not actually using `single_dep` anywhere. I've added the app-services configs in the braindump and checked before and after removing the `single_dep` loader and transforms and it's empty. The code already uses the`multi_dep` logic as it seems. So I think we're good here for now.

JohanLorenzo commented on .taskcluster.yml:
```diff
@@ -1,23 +1,21 @@
 tasks:
   $let:
+    taskgraph:
+      branch: taskgraph-try
```
Nit: Let's put a code comment to explain why `taskgraph-try` is used for now.

JohanLorenzo commented on taskcluster/ci/lint/kind.yml:
```diff
@@ -0,0 +1,28 @@
+  detekt:
+    description: 'detekt'
+    run:
+      command: ./gradlew --no-daemon clean detekt
```
Followup: I should have had noted this on the first pass. Let's use the `gradlew` job type, just like fenix does:
 * Usage: https://github.com/mozilla-mobile/fenix/blob/f8048462b584d04998aec3a3d513e5c6b57697f8/taskcluster/ci/lint/kind.yml#L54-L56
 * Job type definition: https://github.com/mozilla-mobile/fenix/blob/f8048462b584d04998aec3a3d513e5c6b57697f8/taskcluster/fenix_taskgraph/job.py#L70-L82

MihaiTabara replied:
Addressed in https://github.com/mozilla/application-services/pull/3113

MihaiTabara commented:
@eoger This is ready for your review. I've squashed them in a single signed commit ;-) 

@JohanLorenzo I've updated this to use `taskgraph` revision as I landed all depending patches. Once this is reviewed and merge, can you please merge this so that you can keep an eye on it? I'm about to head for PTO and I want to avoid a `hit-and-run` :)

eoger commented on automation/taskcluster/decisionlib.py:
```diff
@@ -223,8 +223,7 @@ def create(self):
         task_id = taskcluster.slugId().decode("utf8")
-        SHARED.schedule_task(task_id, queue_payload)
-        return task_id
+        return task_id, queue_payload
```
Isn't this change going to break the `master`/`release` pushes?

JohanLorenzo replied:
Good call! I guess it was for debugging purpose. Let's change it back to `return task_id`

eoger replied:
The removal of `SHARED.schedule_task(task_id, queue_payload)` is ok?

JohanLorenzo replied:
You're right, there's a second bug there. I didn't see it because I assumed `taskgraph` took care of digesting the tasks decisionlib created and then scheduled them. That's a strategy we applied in Fenix, for instance: https://github.com/mozilla-mobile/fenix/pull/5257. 

I took a closer look and I saw taskgraph is only called on PRs, and decisionlib otherwise. Thus, we do need to keep `SHARED.schedule_task(task_id, queue_payload)` around.

Thank you for spotting this bug! 👍 

That makes me wonder. What are the changes above about? (the ones in `__init__()`) Do we want to revert them too?

JohanLorenzo replied:
Hey @eoger!

Just following up on this issue. Is there anything I can help with? (I'd like to make sure I'm not blocking anything 🙂)

eoger replied:
Sorry I forgot to reply to the earlier comment: yes I think we can revert all the changes in this file!

MihaiTabara replied:
Sorry about this, I forgot it here from during debugging, reverting all the changes to the file.

eoger commented on taskcluster/app_services_taskgraph/build_config.py:
```diff
@@ -0,0 +1,33 @@
+    } for (name, project) in build_config['projects'].items()]
+
+
+@memoize
```
Nit: Shouldn't we memoize `get_components` instead since it applies the same transformation every time?

JohanLorenzo replied:
We definitely can. I guess this solution wasn't implemented because Fenix has slightly different constraints https://github.com/mozilla-mobile/fenix/blob/8be6c77695a7dd5d990999d7bcb5ec09119d3e93/taskcluster/fenix_taskgraph/gradle.py#L33

From what I understand, transforming the data in `get_components()` is not costly and we won't save much by moving `@memoize` there. 

We can address this in a followup, if needed.

MihaiTabara replied:
Added as follow-up.

eoger commented on taskcluster/app_services_taskgraph/loader/__init__.py:
```diff
@@ -0,0 +1,52 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
I don't know what's going on in this file but assume it's fine.

JohanLorenzo replied:
This file defines common functions for loaders. A loader is in charge of taking every task in the kind, applying job-defaults and finding what are the right upstream dependencies. 

If you want to know more about the architecture of taskgraph, you can have a look at: https://johanlorenzo.github.io/blog/2019/10/24/taskgraph-is-now-deployed-to-the-biggest-mozilla-mobile-projects.html. If you don't have much time, I recommend watching the 3-minute-video.

MihaiTabara replied:
Added a comment.

eoger commented on taskcluster/app_services_taskgraph/loader/build_config.py:
```diff
@@ -0,0 +1,25 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
Same here, maybe we should have a quick intro in each file to describe what they do?

JohanLorenzo replied:
This loader is a special one. It's only used by https://github.com/mozilla/application-services/pull/2820/files/39b0c15c030f2aacbf0177b858143b7f8696a94d#diff-037116387fd4f688ffd8855e6f3969d5R5.

It generates a new build task for every component defined in `.buildconfig-android.yml`. It does nothing more.

MihaiTabara replied:
Added an explanatory comment at the beginning of the file.

JohanLorenzo commented on .taskcluster.yml:
```diff
@@ -52,6 +50,90 @@ tasks:
+          else:
+              # GitHub adds "[bot]" to bot usernames and that doesn't validate as email.
+              $if: 'event.sender.login == "bors[bot]"'
+              then: 'skaspari+mozlando@mozilla.com'   # It must match what's in bors.toml
```
At first, I was going to say: I'd prefer to have every `.taskcluster.yml` the same between projects. Sometimes Releng has patches to apply on every taskgraph project and tracking differences becomes a difficult task if differences are too big.

That said, I'm certain if one day, you enable bors, you won't likely use @pocmo's email address. Thus, I'm okay to remove these bits of config. 

I'll take care of it in a followup.

MihaiTabara replied:
I removed the bors section entirely for now.

eoger commented:
This removed support for https://github.com/mozilla/application-services/pull/2923, would it be possible to put it back?
